lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3391,6 +3394,14 @@ namespace {
         func->setAccess(AccessLevel::Public);
       }
 
+      if (!isa<clang::CXXConstructorDecl>(decl) && !importedType) {
+        if (!Impl.importFunctionReturnType(decl, result->getDeclContext())) {
```

Comment:
We intentionally don't eagerly import some class template specializations. These really need to be lazily evaluated to respect C++ semantics. You are allowed to have a completely broken class template specialization, as long as the method is never called.

Reply:
We are currently handling this logic on line 3304/3307: `if (!bodyParams)`. We use body params, because we sometimes want to import the return type lazily. For better or worse, this is an invariant that we've had in the importer, we should change it with caution.

Reply:
To elaborate on the first comment, this is the case I'm talking about, where we may violate C++ "semantics" if we try to do this too eagerly:

```
template<class T> struct Borked { static_assert(__is_same(T, char), "I'm broken!"); };

struct HasMethod { Borked<int> test(); };

void test(HasMethod hm) {
    // As long as this is never called, we're fine.
    // hm.test();
}
```

In other words, we cannot import `test`'s return type eagerly here.